syntax = "proto3";

/*
 This container message is send from Lo Level to Hi level and may contain any allowed message in that direction.
*/
message LoToHi {
 oneof payload {
  // Common Packets
  KeepAliveLo keep_alive = 3;
  PowerMeter power_meter = 4;
  ResetDone reset_done = 7;
  // Packets only available in control_mode = 2 (low level control)
  Event event = 6;
  Telemetry telemetry = 8;
 }
}

message Event {
  enum InterfaceEvent {
    CAR_PLUGGED_IN = 0;
    CAR_REQUESTED_POWER = 1;
    POWER_ON = 2;
    POWER_OFF = 3;
    CAR_REQUESTED_STOP_POWER = 4;
    CAR_UNPLUGGED = 5;
    ERROR_E = 6;
    ERROR_DF = 7;
    ERROR_RELAIS = 8;
    ERROR_RCD = 9;
    ERROR_VENTILATION_NOT_AVAILABLE = 10;
    ERROR_OVER_CURRENT = 11;
    ENTER_BCD = 12;
    LEAVE_BCD = 13;
    PERMANENT_FAULT = 14;
    EVSE_REPLUG_STARTED = 15;
    EVSE_REPLUG_FINISHED = 16;
  }
  InterfaceEvent type = 1;
}

message PowerMeter {
  float voltage = 2;
}

message Telemetry {
  float cp_hi = 1;
  float cp_lo = 2;
  float pwm_dc = 3;
  float relais_on = 4;
}

message KeepAliveLo {
  uint32 time_stamp = 1;
  uint32 hw_type = 2;
  uint32 hw_revision = 3;
  uint32 protocol_version_major = 4;
  uint32 protocol_version_minor = 5;
  string sw_version_string = 6;
}

message ResetDone {
}

